From d18b5ed9d0faf8666d0e3c9d2eddcf373b84c72b Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sat, 9 Dec 2017 15:01:28 +0000
Subject: [PATCH 4/6] dependencies: Work around ugly bug in eopkg with filesdb

When updating and having files conflicts by swapping some files around,
eopkg filesdb becomes distinctly less useful by "forgetting" the keys in
the filesdb and can only be remedied by an `eopkg rdb` call.

In those instances we'll fallback to a slow search and try to find the
guy in question.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 ypkg2/dependencies.py | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/ypkg2/dependencies.py b/ypkg2/dependencies.py
index caa187e..8d743e5 100644
--- a/ypkg2/dependencies.py
+++ b/ypkg2/dependencies.py
@@ -63,6 +63,8 @@ class DependencyResolver:
 
     kernel_cache = dict()
 
+    deadends = dict()
+
     # Cached from packagedb
     pkgConfigs = None
     pkgConfigs32 = None
@@ -70,9 +72,18 @@ class DependencyResolver:
     def search_file(self, fname):
         if fname[0] == '/':
             fname = fname[1:]
-        if not self.fdb.has_file(fname):
+        if fname in self.deadends:
             return None
-        return self.fdb.get_file(fname)
+        if self.fdb.has_file(fname):
+            return self.fdb.get_file(fname)
+        # Nasty file conflict crap happened on update and the filesdb
+        # is now inconsistent ..
+        ret = self.fdb.search_file(fname)
+        if len(ret) == 1:
+            return ret[0]
+        # Just blacklist further lookups here
+        self.deadends[fname] = 0
+        return None
 
     def __init__(self):
         """ Allows us to do look ups on all packages """
-- 
2.16.1

